AppKit/CPTextView/CPTextView.j

```diff
@@ -0,0 +1,1890 @@
+        style.whiteSpace = "pre";
+        style.backgroundColor = "black";
+        _caretDOM.style.width = "1px";
+        _DOMElement.appendChild(_caretDOM);
+    }
+
+    _caretDOM.style.left = (aRect.origin.x) + "px";
+    _caretDOM.style.top = (aRect.origin.y) + "px";
+    _caretDOM.style.height = (aRect.size.height) + "px";
+    _caretDOM.style.visibility = flag ? "visible" : "hidden";
+#endif
+}
+
+- (id) _createSelectionSpanForRect:(CPRect)aRect andColor:(CPColor)aColor
+{
+#if PLATFORM(DOM)
+    var ret = document.createElement("span");
+    ret.style.position = "absolute";
+    ret.style.visibility = "visible";
+    ret.style.padding = "0px";
+    ret.style.margin = "0px";
+    ret.style.whiteSpace = "pre";
+    ret.style.backgroundColor = [aColor cssString];
+
+    ret.style.width = (aRect.size.width)+"px";
+    ret.style.left = (aRect.origin.x) + "px";
+    ret.style.top = (aRect.origin.y) + "px";
+    ret.style.height = (aRect.size.height) + "px";
+    ret.style.zIndex = -1000;
+    ret.oncontextmenu = ret.onmousedown = ret.onselectstart = function () { return false; };
+    return ret;
+#else
+    return nil;
+#endif
+}
+
+- (void)drawRect:(CGRect)aRect
+{
+#if PLATFORM(DOM)
+    var range = [_layoutManager glyphRangeForBoundingRect:aRect inTextContainer:_textContainer];
+
+    if (_selectionSpans)
+    {
+        for (var i = 0; i < _selectionSpans.length; i++)
+        {
+            _DOMElement.removeChild(_selectionSpans[i]);
+        }
+    }
+    _selectionSpans = [];
+
+    if (_selectionRange.length)
+    {
+        var rects = [_layoutManager rectArrayForCharacterRange:_selectionRange
+                                  withinSelectedCharacterRange:_selectionRange
+                                               inTextContainer:_textContainer
+                                                     rectCount:nil],
+            effectiveSelectionColor = [self _isFocused] ? [_selectedTextAttributes objectForKey:CPBackgroundColorAttributeName] : [CPColor _selectedTextBackgroundColorUnfocussed],
+            lengthRect = rects.length;
+
+
+        for (var i = 0; i < lengthRect; i++)
+        {
+            rects[i].origin.x += _textContainerOrigin.x;
+            rects[i].origin.y += _textContainerOrigin.y;
+
+            var newSpan = [self _createSelectionSpanForRect:rects[i] andColor:effectiveSelectionColor];
+            _selectionSpans.push(newSpan);
+            _DOMElement.appendChild(newSpan);
+        }
+    }
+
+    if (range.length)
+        [_layoutManager drawGlyphsForGlyphRange:range atPoint:_textContainerOrigin];
+
+    if ([self shouldDrawInsertionPoint])
+    {
+        [self updateInsertionPointStateAndRestartTimer:NO];
+        [self drawInsertionPointInRect:_caretRect color:_insertionPointColor turnedOn:_drawCaret];
+    }
+    else // <!> FIXME: breaks DOM abstraction, but i did get it working otherwise
+    {
+        if (_caretDOM)
+            _caretDOM.style.visibility = "hidden";
```

Comment:
With a private class we could avoid to have this kind of line of code, but something more objective-j